actors/miner/src/partition_state.rs

```diff
@@ -822,6 +822,43 @@ impl Partition {
         self.validate_bf_state()?;
         Ok(())
     }
+
+    // ajust epoch info of expirations_epochs and early_terminated within `Partition`
+    pub fn adjust_for_move<BS: Blockstore>(
+        &mut self,
+        store: &BS,
+        dest_quant: &QuantSpec,
+    ) -> anyhow::Result<()> {
+        let orig_expirations_epochs: Array<ExpirationSet, _> =
+            Array::load(&self.expirations_epochs, store)?;
+
+        let orig_early_terminations: Array<BitField, _> =
+            Array::load(&self.early_terminated, store)?;
+
+        let mut dest_expirations_epochs = Array::<ExpirationSet, BS>::new_with_bit_width(
+            store,
+            PARTITION_EXPIRATION_AMT_BITWIDTH,
+        );
+        let mut dest_early_terminations = Array::<BitField, BS>::new_with_bit_width(
+            store,
+            PARTITION_EARLY_TERMINATION_ARRAY_AMT_BITWIDTH,
+        );
+
+        orig_expirations_epochs.for_each(|from_epoch, expire_set| {
+            dest_expirations_epochs
+                .set(dest_quant.quantize_up(from_epoch as ChainEpoch) as u64, expire_set.clone())?;
+            Ok(())
+        })?;
+        orig_early_terminations.for_each(|from_epoch, bitfield| {
+            dest_early_terminations
+                .set(dest_quant.quantize_up(from_epoch as ChainEpoch) as u64, bitfield.clone())?;
+            Ok(())
+        })?;
```

Comment:
This will make it possible to indefinitely extend sector expiration by rotating them around the clock.

While it would lead to a larger patch and a bigger upgrade, the best (IMO) way to do this would be to change how we quantize. Instead of using epochs as the keys, we'd use "days" as keys. I.e., we'd:

1. Take the expiration.
2. Divide by `EPOCHS_PER_DAY`, rounded up.

(approximately)

That way we wouldn't need to do this rescheduling at all (and these expiration queues will be more efficiently packed).

cc @anorth for feedback in case I'm missing something.

Reply:
Alternatively, we could re-quantize based on the sector infos, but that's likely going to be _super_ expensive.

Reply:
@Stebalien as there's already a `move_partitions` function `Actor::move_partitions`, and this one is related with two `Deadline`, so I defined it inside `Deadlines` as an associate function.. Where do you recommend to put it as a free function?

Reply:
@Stebalien Do you mean a migration of all existing `quantize_up/quantize_down` results into using "days" as keys?  Or only the `expirations_epochs/early_terminated` involved here? The former is going to be a big change actually.

I'll also wait @anorth for feedback because we once talked about this inaccuracy in the previous threads [here](https://github.com/filecoin-project/builtin-actors/pull/1326#discussion_r1270155493).

Reply:
It would be all of them, unfortunately.

But to be clear, the previous conversation considered _one_ move. Here, I'm considering _repeated_ moves which would let you move an expiration arbitrarily far into the future (which is a hard-no from a security perspective).

Reply:
Moving the conversation here from DMs: An alternative approach is to:

1. Quantize up when moving from a lower deadline to a higher deadline (e.g., 0 -> 5).
2. Quantize down when moving from a higher deadline to a lower deadline (e.g., 30 -> 1).

Unless I'm mistaken, this should prevent "ratcheting" up an expiration. It will, however, let sectors expire up to 24 hours early, which isn't great. But even the solution I proposed above would have had that same issue.

Reply:
@Stebalien I think it's an ingenious idea! When moving from lower to higher, the result epoch will be higher, but within 24 hours;When moving from higher to lower, the result epoch will be lower, but still within 24 hours. 

But I've a concern, for `expirations_epochs`, will it break some invariant if the result expire epoch is lower than actual expire epoch? And for the early_terminated , it seems the `QuantSpec` used is `NO_QUANTIZATION`, maybe we should also use `NO_QUANTIZATION`  here?

Reply:
In terms of impact, I'm not sure. I don't _think_ it'll cause a problem, but you'd need to look through the code.

E.g., it _could_ cause issues with deals... But there's usually quite a bit of "fuzz" room in expiration times anyways.

In terms of `NO_QUANTIZATION`... we _could_ do that (I think) and it would lead to the sector expiring on the _next_ deadline. However, it would also bloat state and would lead to some extra work in cron, so I'm not a fan of that solution.

Reply:
I think I understand the issue and its possibility. I concur that re-quantizing from sector infos will be too expensive to consider a decent solution.

Quantizing daily has merit and is worth exploring. I interpret this to mean quantizing upward to a miner's _proving period end_, after whatever deadline end the epoch would otherwise have been. This would have a notable downside of concentrating sector expiration processing (in cron today) into one batch daily, rather than 48 batches. We might be able to do it without migrating all existing queues to this form: in an ugly way there are just two places to check. But this would require some fairly gnarly testing and generally adds complexity.

I do think that expiring a sector before its scheduled epoch is going to cause problems. With the market actor is one potential problem: it could result in deals being terminated rather than successfully completed. The assumption that sector expiration is processed only after the declared epoch is probably elsewhere too.

I don't have a solution yet, let's keep exploring.

Reply:
> Quantizing daily has merit and is worth exploring. I interpret this to mean quantizing upward to a miner's proving period end, after whatever deadline end the epoch would otherwise have been. This would have a notable downside of concentrating sector expiration processing (in cron today) into one batch daily, rather than 48 batches. We might be able to do it without migrating all existing queues to this form: in an ugly way there are just two places to check. But this would require some fairly gnarly testing and generally adds complexity.

I meant something slightly different:

- Right now we quantize by rounding up to the next time the deadline to which the sector was assigned will be processed.
- Instead, I'm proposing that we simply divide all expirations by 2880 (epochs in a day), then round up to the next integer (call this an expiration _slot_). When processing expirations, we'd divide the current epoch by 2880, round down, then process all expiration slots  up to and including the "slot" for the current epoch.

Moving a partition between deadlines would change when the sectors expire (possibly moving the expiration up to 24hr into the future), but we wouldn't even need to rewrite the expiration queue because the slots would stay the same.

This would also give us more compact expiration queues (which may actually hurt performance... now that I think about it).